bench/scalac212/src/main/scala/rsc/bench/ScalacCompile.scala

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2017 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0 (see LICENSE.md).
+package rsc.bench
+
+import java.nio.file._
+import java.util.concurrent.TimeUnit
+import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.annotations.Mode._
+import scala.tools.nsc._
+import scala.tools.nsc.reporters._
+import rsc.bench.ScalacCompile._
+
+object ScalacCompile {
+  @State(Scope.Benchmark)
+  class BenchmarkState extends FileFixtures
+}
+
+trait ScalacCompile {
+  def runImpl(bs: BenchmarkState): Unit = {
+    val settings = new Settings
+    settings.outdir.value = Files.createTempDirectory("scalac_").toString
+    settings.usejavacp.value = true
+    val reporter = new StoreReporter
+    val global = Global(settings, reporter)
+    val run = new global.Run
+    run.compile(bs.re2sScalacFiles.map(_.toString))
+    if (reporter.hasErrors) {
+      reporter.infos.foreach(println)
+      sys.error("compile failed")
+    }
+  }
+}
+
+@BenchmarkMode(Array(SingleShotTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Fork(value = 128, jvmArgs = Array("-Xms2G", "-Xmx2G"))
+class ColdScalacCompile extends ScalacCompile {
+  @Benchmark
+  def run(bs: BenchmarkState): Unit = {
+    runImpl(bs)
+  }
+}
+
+@BenchmarkMode(Array(SampleTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 10, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 10, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 1, jvmArgs = Array("-Xms2G", "-Xmx2G"))
+class WarmScalacCompile extends ScalacCompile {
+  @Benchmark
+  def run(bs: BenchmarkState): Unit = {
+    runImpl(bs)
+  }
+}
+
+@BenchmarkMode(Array(SampleTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 3, jvmArgs = Array("-Xms2G", "-Xmx2G"))
+class HotScalacCompile extends ScalacCompile {
+  @Benchmark
+  def run(bs: BenchmarkState): Unit = {
+    runImpl(bs)
+  }
+}
```

Comment:
@retronym I've been meaning to compare notes wrt. scala/compiler-benchmark right after we opensourced Rsc, but forgot about that in the post-release rush. Looks like we have another good opportunity to chat.

Just like you guys, we found it useful to split benchmarks into `cold`, `warm` and `hot`. In our benchmarks, `cold` and `hot` cover two important use cases - single-shot CLI runs and resident compilations. Moreover, we use `warm` to get a quick & dirty approximation of `hot` during interactive development.

Unfortunately, because of the peculiar configuration style of JMH, we haven't found a way to abstract these notions, so we have to create endless bunches of `ColdXxxYyy`, `WarmXxxYyy` and `HotXxxYyy` classes with the same JMH annotations repeated over and over again. 

I guess you guys faced the same problem in scala/compiler-benchmark. Do you maybe have some ideas how to cut down on duplication?